### DISCUSSION TITLE
Oqtane runtime configuration

### DISCUSSION THREAD

**Discussion by bitzi** [oqtane/oqtane.framework#1223] [Q&A]
Hi,

is it possible for Oqtane to support Blazor's server-mode "ServerPrerendered"? 
I think that with Oqtane's current supported runtimes "Server" and "WebAssembly" only, SEO it's going to be a big issue for websites built with Oqtane.

**Comment by sbwalker** [comment]
Blazor's "render-mode" attribute on the root component tag is unrelated to Oqtane's Runtime property which can be set in appsettings.json. Currently "render-mode" is set to "Server" and cannot be modified through the framework - it needs to be manually changed in _Host.cshtml ( which is obviously not recommended as you would be modifying the core code ). It would require some effort to support other render-modes, as each option has its own requirements.

It is important to note that Oqtane is not a CMS - it is a modular application framework ( this is highlighted in the Oqtane Philosophy: https://www.oqtane.org/Resources/Blog/PostId/538/oqtane-philosophy ). The framework is optimized for developing SaaS solutions - not publishing content-focused websites.

**Reply by sbwalker** [reply]
So I recently did some testing and changed the render-mode in _Host.cshtml from
```
    <app>
        <component type="typeof(Oqtane.App)" render-mode="Server" />
    </app>
```
to
```
    <app>
        <component type="typeof(Oqtane.App)" render-mode="ServerPrerendered" />
    </app>
```

and the framework runs just fine without any errors or behavioral issues. However it is still not 100% clear to me what is happening behind the scenes. I used my browser developer tools to review the HTML output produced with both settings... and they are identical. This does not seem to be consistent with the render-mode documentation:

Server - Renders a marker for a Blazor server-side application. This doesn't include any output from the component. When the user-agent starts, it uses this marker to bootstrap a blazor application.

ServerPrerendered - Renders the component into static HTML and includes a marker for a Blazor server-side application. When the user-agent starts, it uses this marker to bootstrap a blazor application.

Perhaps the browser developer tools are not able to capture the rendered output under each scenario? I am wondering how to effectively test this feature.

**Reply by bitzi** [reply]
Interesting .... I tried the same, I set render-mode="ServerPrerendered" and it worked: I looked at the page source in the browser and I could see the html prerendered on the server. I switched back to render-mode="Server" and no html prerendered, as expected.
So, it works! I used the master branch and Chrome browser.

**Reply by bitzi** [reply]
I cannot say the same for WebAssemblyPrerendered: in the page source it shows  

<script src="_framework/blazor.server.js"></script> 
instead of 
<script src="_framework/blazor.webassembly.js"></script>

although the html is prerendered but the site menu it looks bad, like the css is not there.

I used Runtime:"Client" in the appsettings.json and render-mode="WebAssemblyPrerendered" in _Host.html
